From 3f90105fb924c8ad7bc3f0d8d7e4aa170ee9306d Mon Sep 17 00:00:00 2001
From: Eric Sandeen <sandeen@redhat.com>
Date: Mon, 4 Aug 2014 19:04:56 -0400
Subject: [fs] btrfs: assert send doesn't attempt to start transactions

Message-id: <1407179109-13714-410-git-send-email-sandeen@redhat.com>
Patchwork-id: 87279
O-Subject: [RHEL7.1 PATCH 409/422] Btrfs: assert send doesn't attempt to start transactions
Bugzilla: 1071179
RH-Acked-by: Zach Brown <zab@redhat.com>

From: Filipe Manana <fdmanana@gmail.com>

commit 46c4e71e9b02a649c722f06569f5b6575da02dba
Author: Filipe Manana <fdmanana@gmail.com>
Date:   Tue Jun 24 17:48:28 2014 +0100

    Btrfs: assert send doesn't attempt to start transactions

    When starting a transaction just assert that current->journal_info
    doesn't contain a send transaction stub, since send isn't supposed
    to start transactions and when it finishes (either successfully or
    not) it's supposed to set current->journal_info to NULL.

    This is motivated by the change titled:

        Btrfs: fix crash when starting transaction

    Signed-off-by: Filipe David Borba Manana <fdmanana@gmail.com>
    Signed-off-by: Chris Mason <clm@fb.com>
Signed-off-by: Eric Sandeen <sandeen@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/btrfs/transaction.c b/fs/btrfs/transaction.c
index ac984a3..568a320 100644
--- a/fs/btrfs/transaction.c
+++ b/fs/btrfs/transaction.c
@@ -386,11 +386,13 @@ start_transaction(struct btrfs_root *root, u64 num_items, unsigned int type,
  bool reloc_reserved = false;
  int ret;
 
+ /* Send isn't supposed to start transactions. */
+ ASSERT(current->journal_info != (void *)BTRFS_SEND_TRANS_STUB);
+
  if (test_bit(BTRFS_FS_STATE_ERROR, &root->fs_info->fs_state))
   return ERR_PTR(-EROFS);
 
- if (current->journal_info &&
-     current->journal_info != (void *)BTRFS_SEND_TRANS_STUB) {
+ if (current->journal_info) {
   WARN_ON(type & TRANS_EXTWRITERS);
   h = current->journal_info;
   h->use_count++;
-- 
1.7.1